source/slang/slang-capability.cpp

```diff
@@ -1331,4 +863,211 @@ void printDiagnosticArg(StringBuilder& sb, CapabilityName name)
+
+    testCapSetA = CapabilitySet(CapabilityName::TEST_JOIN_2A);
+    testCapSetB = CapabilitySet(CapabilityName::TEST_JOIN_2B);
+    testCapSetA.join(testCapSetB);
+
+    CHECK_CAPS(TEST_findTargetCapSet(testCapSetA, CapabilityAtom::hlsl));
+    CHECK_CAPS(TEST_targetCapSetWithSpecificSetInStage(testCapSetA, CapabilityAtom::hlsl, CapabilityAtom::vertex,
+        { CapabilityAtom::textualTarget, CapabilityAtom::hlsl, CapabilityAtom::vertex,
+        CapabilityAtom::_sm_4_0, CapabilityAtom::_sm_4_1 }));
+    
+    // ------------------------------------------------------------
+
+    testCapSetA = CapabilitySet(CapabilityName::TEST_JOIN_3A);
+    testCapSetB = CapabilitySet(CapabilityName::TEST_JOIN_3B);
+    testCapSetA.join(testCapSetB);
+
+    CHECK_CAPS((int)!TEST_findTargetCapSet(testCapSetA, CapabilityAtom::spirv_1_0));
+    CHECK_CAPS(TEST_findTargetCapSet(testCapSetA, CapabilityAtom::glsl));
+    CHECK_CAPS((int)!TEST_findTargetStage(testCapSetA, CapabilityAtom::glsl, CapabilityAtom::raygen));
+    CHECK_CAPS(TEST_targetCapSetWithSpecificSetInStage(testCapSetA, CapabilityAtom::glsl, CapabilityAtom::fragment,
+        { CapabilityAtom::textualTarget, CapabilityAtom::glsl, CapabilityAtom::fragment,
+        CapabilityAtom::_GLSL_130, CapabilityAtom::_GLSL_140 }));
+    CHECK_CAPS(TEST_targetCapSetWithSpecificSetInStage(testCapSetA, CapabilityAtom::glsl, CapabilityAtom::vertex,
+        { CapabilityAtom::textualTarget, CapabilityAtom::glsl, CapabilityAtom::vertex,
+        CapabilityAtom::_GLSL_130, CapabilityAtom::_GLSL_140 }));
+    CHECK_CAPS(TEST_targetCapSetWithSpecificSetInStage(testCapSetA, CapabilityAtom::hlsl, CapabilityAtom::fragment,
+        { CapabilityAtom::textualTarget, CapabilityAtom::hlsl, CapabilityAtom::fragment,
+        CapabilityAtom::_sm_4_0, CapabilityAtom::_sm_4_1 }));
+    CHECK_CAPS(TEST_targetCapSetWithSpecificSetInStage(testCapSetA, CapabilityAtom::hlsl, CapabilityAtom::vertex,
+        { CapabilityAtom::textualTarget, CapabilityAtom::hlsl, CapabilityAtom::vertex,
+        CapabilityAtom::_sm_4_0 }));
+
+    // ------------------------------------------------------------
+
+    testCapSetA = CapabilitySet(CapabilityName::TEST_JOIN_4A);
+    testCapSetB = CapabilitySet(CapabilityName::TEST_JOIN_4B);
+    testCapSetA.join(testCapSetB);
+
+    CHECK_CAPS(TEST_findTargetCapSet(testCapSetA, CapabilityAtom::glsl));
+    CHECK_CAPS(TEST_targetCapSetWithSpecificSetInStage(testCapSetA, CapabilityAtom::glsl, CapabilityAtom::fragment,
+        { CapabilityAtom::textualTarget, CapabilityAtom::glsl, CapabilityAtom::fragment,
+        CapabilityAtom::_GLSL_130, CapabilityAtom::_GLSL_140, CapabilityAtom::_GLSL_150, CapabilityAtom::_GL_EXT_texture_query_lod, CapabilityAtom::_GL_EXT_texture_shadow_lod }));
+
+    // ------------------------------------------------------------
+
+
+}
+
+void TEST_CapabilitySet()
+{
+    TEST_CapabilitySet_addAtom();
+    TEST_CapabilitySet_join();
+}
```

Comment:
Where is this function gets called from?

Reply:
This is some testing functions I was asked to make.
I was asked to macro out the code and keep it unused until we expand the Unit test system to target the capabilities system.

Reply:
I like the unit testing idea.
Where is this function called from tho?
You said some is "macro out"? What do you mean by that?

Reply:
I was advised to not remove the testing code

Reply:
Where is this function called from?
I am not saying you should remove it.
In fact, I like the unit testing.
I am trying to understand how you initiate the unit test.

Reply:
these functions are all mutually and functionally exclusive tests

call any one of them to test.